File: Traditional-Algorithms/LeetCode34.cpp
//for循环，速度慢，最差情况下需要检索整个数组
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans;
        for(int i = 0 ; i < nums.size(); i++){
            if(nums[i] == target){
                ans.push_back(i);
            }
            if(nums[i] > target){
                break;
            }
        }
        if(ans.empty()){
            return {-1,-1};
        }
        if(ans.size() == 1){
            return {ans[0], ans[0]};
        }
        return {ans[0],ans[ans.size()-1]};
    }
};

File: Traditional-Algorithms/golang/leetcode34.go
func searchRange(nums []int, target int) []int {
    length := len(nums)
    if length == 0{
        return []int{-1, -1}
    }
    first := findFirstElem(nums, target)
    end := findLastElem(nums, target)
    result := []int{first, end}
    return result
}

func findFirstElem(nums []int, target int) int{
    length := len(nums)
    left := 0
    right := length -1
    for;left < right;{
        mid := left + (right - left) / 2
        if nums[mid] < target{
            left = mid + 1
        }else{
            right = mid
        }
    }
    if nums[left] == target{
        return left
    }
    return -1
}

func findLastElem(nums []int, target int) int{
    length := len(nums)
    left := 0
    right := length - 1
    for;left < right;{
        mid := left + (right - left + 1)/2
        if nums[mid] > target{
            right = mid -1
        }else{
            left = mid
        }
    }
    if nums[left] == target{
        return left
    }
    return -1
}